File: matlab/rigid_registration.m
%% Rigid Registration
% Cameron Wolfe 03/26/2024
%
% Performs a rigid registration between the model and aurora frame,
% providing T_aurora_2_model.  Then performas a rigid registration to
% find T_coil_2_tip.  RMSE is returned for all transforms gotten from
% data.  Additionally, T_tip_2_model is calculated and displayed, but
% not returned.
%
% INPUTS:
%  - REG_FILE: (string) Aurora data file (0A: pen, 0B: coil)
%  - PEN_FILE: (string) Pen pivot_cal file
%  - output_files: (boolean) If output files are desired
%
% OUTPUTS:
%  - T_aurora_2_model: (4x4 double) Aurora to model frame transformation matrix
%  - T_tip_2_coil: (4x4 double) tip to coil frame transformation matrix
%  - rmse: (struct) Contains RMS error for various transformations

function [T_aurora_2_model, T_tip_2_coil, T_tip_2_model, rmse] = rigid_registration(REG_FILE, PEN_FILE, output_files, repetitions)

%% Input parsing
if nargin == 2
    output_files = false;
    repetitions = 1;
end
if nargin == 3
    repetitions = 1;
end

%% Setup
% Truth filenames
SW_MODEL_POS_FILE = "../tools/model_registration_points_in_sw";
SW_TIP_POS_FILE = "../tools/all_tip_registration_points_in_sw";

T_SW_2_MODEL_FILE = "../tools/T_sw_2_model";
T_SW_2_TIP_FILE = "../tools/T_sw_2_tip";

% RMSE error struct initialization
rmse = struct();

%% File inputs
model_reg_truth_in_sw = readmatrix(SW_MODEL_POS_FILE);
tip_reg_truth_in_sw = readmatrix(SW_TIP_POS_FILE);

T_sw_2_model = readmatrix(T_SW_2_MODEL_FILE);
T_sw_2_tip = readmatrix(T_SW_2_TIP_FILE);

penprobe = readmatrix(PEN_FILE);
reg_table = readtable(REG_FILE);

%% Get truth registration positions
model_reg_truth_in_model = T_mult(T_sw_2_model, model_reg_truth_in_sw);
tip_reg_truth_in_tip = T_mult(T_sw_2_tip, tip_reg_truth_in_sw);

% If multiple repetitions of a point are collected, use all of them in reg
temp_model = zeros(3, length(model_reg_truth_in_model) * repetitions);
temp_tip = zeros(3, length(tip_reg_truth_in_tip) * repetitions);

for i=1:repetitions
    for j=1:length(model_reg_truth_in_model)
        temp_model(:, (j - 1) * repetitions + i) = model_reg_truth_in_model(:, j);
    end
    for j=1:length(tip_reg_truth_in_tip)
        temp_tip(:, (j - 1) * repetitions + i) = tip_reg_truth_in_tip(:, j);
    end
end

model_reg_truth_in_model = temp_model;
tip_reg_truth_in_tip = temp_tip;

%% Extracting and processing aurora data
pen_mat = table2array(reg_table(strcmp('0A', reg_table{:, 3}), 4:10));
tip_mat = table2array(reg_table(strcmp('0B', reg_table{:, 3}), 4:10));

% Use penprobe to get pen tip position
pen_positions = nan(3, size(pen_mat,1));
for i=1:size(pen_mat, 1)
    R = quat2matrix(pen_mat(i, 1:4));
    pen_positions(:, i) = pen_mat(i, 5:7)' + R * penprobe';
end

% Split into model registration points and tip registration points
model_reg_meas_in_aurora = pen_positions(:, 1:size(model_reg_truth_in_model, 2));
tip_reg_meas_in_aurora = pen_positions(:, size(model_reg_truth_in_model, 2)+1:end);

%% SVD rigid registrations
[~, T_aurora_2_model, rmse.aurora_2_model] = rigid_align_svd(model_reg_meas_in_aurora, model_reg_truth_in_model);

T_aurora_2_model

[~, T_aurora_2_tip, rmse.aurora_2_tip] = rigid_align_svd(tip_reg_meas_in_aurora, tip_reg_truth_in_tip);

T_aurora_2_tip

%% Coil to Aurora registration
mean_tip_quat = quat_mean(tip_mat(:,1:4));
mean_tip_pos = mean(tip_mat(:, 5:end));

T_coil_2_aurora = [[quat2matrix(mean_tip_quat), mean_tip_pos']; [0 0 0 1]]
rmse.coil_2_aurora = sqrt(mean((tip_mat(:,5:end) - mean_tip_pos).^2, 'all'));

%% Final transforms
T_tip_2_coil = T_coil_2_aurora^-1 * T_aurora_2_tip^-1
rmse.tip_2_coil_est = sqrt(rmse.coil_2_aurora^2 + rmse.aurora_2_tip^2);
T_tip_2_model = T_aurora_2_model * T_coil_2_aurora * T_tip_2_coil
rmse.tip_2_model_est = sqrt(rmse.aurora_2_model^2 + rmse.coil_2_aurora^2 + rmse.tip_2_coil_est^2);

T_tip_2_model_truth = T_sw_2_model * T_sw_2_tip^-1

%% File outputs
if output_files == true
    % Output filenames
    T_AURORA_2_MODEL_FILE = "../tools/T_aurora_2_model";
    T_TIP_2_COIL_FILE = "../tools/T_tip_2_coil";
    
    writematrix(T_aurora_2_model, T_AURORA_2_MODEL_FILE);
    system(("mv " + T_AURORA_2_MODEL_FILE + ".txt " + T_AURORA_2_MODEL_FILE)); % Get rid of .txt
    writematrix(T_tip_2_coil, T_TIP_2_COIL_FILE);
    system(("mv " + T_TIP_2_COIL_FILE + ".txt " + T_TIP_2_COIL_FILE)); % Get rid of .txt
end
end

File: python/rigid_registration.py
#!/bin/python3
from pathlib import Path
import numpy as np
import matplotlib.pyplot as plt
import kinematics
import utils_data

# Parameters
REG_FILE = "../data/regs/reg_07_15_24c.csv"
TIP_FILE = "../tools/penprobe_07_03_24i"
output = True
num_repetitions = 5

# Setup
# Truth filenames
SW_MODEL_POS_FILE = "../tools/12_model_registration_points_in_sw"
SW_TIP_POS_FILE = "../tools/all_tip_registration_points_in_sw"
T_SW_2_MODEL_FILE = "../tools/T_sw_2_model"
T_SW_2_TIP_FILE = "../tools/T_sw_2_tip"

# FRE (fiducial regisitration error) dictionary initialization
fre = {}

# File inputs
model_truth_in_sw = np.loadtxt(Path(SW_MODEL_POS_FILE), delimiter=",")
tip_truth_in_sw = np.loadtxt(Path(SW_TIP_POS_FILE), delimiter=",")
T_sw_2_model = np.loadtxt(Path(T_SW_2_MODEL_FILE), delimiter=",")
T_sw_2_tip = np.loadtxt(Path(T_SW_2_TIP_FILE), delimiter=",")
penprobe = np.loadtxt(Path(TIP_FILE), delimiter=",")

aurora_transforms = utils_data.parse_aurora_csv(REG_FILE)

model_truth_in_model = kinematics.Tmult(T_sw_2_model, model_truth_in_sw)
tip_truth_in_tip = kinematics.Tmult(T_sw_2_tip, tip_truth_in_sw)
print(tip_truth_in_tip)

temp_model = np.zeros((3, model_truth_in_model.shape[1] * num_repetitions))
temp_tip = np.zeros((3, tip_truth_in_tip.shape[1] * num_repetitions))

for i in range(num_repetitions):
    for j in range(model_truth_in_model.shape[1]):
        temp_model[:, j * num_repetitions + i] = model_truth_in_model[:, j]
    for j in range(tip_truth_in_tip.shape[1]):
        temp_tip[:, j * num_repetitions + i] = tip_truth_in_tip[:, j]

model_truth_in_model = temp_model
tip_truth_in_tip = temp_tip

num_model_points = model_truth_in_model.shape[1]
num_tip_points = tip_truth_in_tip.shape[1]

# Get truth positions in model and tip frames
# Get penprobe positions
meas_in_aurora = kinematics.penprobe_transform(penprobe, aurora_transforms["0A"])

model_meas_in_aurora = meas_in_aurora[:, :num_model_points]
tip_meas_in_aurora = meas_in_aurora[
    :, num_model_points : num_model_points + num_tip_points
]

fig = plt.figure()
ax = fig.add_subplot(projection="3d")
ax.plot(
    model_meas_in_aurora[0, :], model_meas_in_aurora[1, :], model_meas_in_aurora[2, :]
)
ax.plot(tip_meas_in_aurora[0, :], tip_meas_in_aurora[1, :], tip_meas_in_aurora[2, :])
plt.show()

# Perform registrations
model_meas_in_model, T_aurora_2_model, fre["aurora_2_model"] = (
    kinematics.rigid_align_svd(model_meas_in_aurora, model_truth_in_model)
)
tip_meas_in_tip, T_aurora_2_tip, fre["aurora_2_tip"] = kinematics.rigid_align_svd(
    tip_meas_in_aurora, tip_truth_in_tip
)

# Get average coil transform
coil_qs = np.concatenate(
    [transform[0].reshape((4, 1)) for transform in aurora_transforms["0B"]], axis=1
)
coil_ts = np.concatenate(
    [transform[1].reshape((3, 1)) for transform in aurora_transforms["0B"]], axis=1
)

q_mean = coil_qs.mean(axis=1)
t_mean = coil_ts.mean(axis=1)

T_coil_2_aurora = np.identity(4)
T_coil_2_aurora[0:3, 0:3] = kinematics.quat_2_dcm(q_mean)
T_coil_2_aurora[0:3, 3] = t_mean.flatten()

T_tip_2_coil = np.linalg.inv(T_coil_2_aurora) @ np.linalg.inv(T_aurora_2_tip)
np.set_printoptions(suppress=True)
print("T_aurora_2_model")
print(T_aurora_2_model)
print("T_aurora_2_tip")
print(T_aurora_2_tip)
print("T_coil_2_aurora")
print(T_coil_2_aurora)
print("T_tip_2_coil")
print(T_tip_2_coil)
print("T_tip_2_model")
print(T_aurora_2_model @ np.linalg.inv(T_aurora_2_tip))
print(fre)

fle = {}

for key, value in fre.items():
    fle[key] = np.sqrt(value**2 * (1 - 2 / (12 * num_repetitions)))

tre = {}

model_cov = model_truth_in_model @ np.transpose(model_truth_in_model)
model_tre_factor = 1 / (12 * num_repetitions) + 1 / 3 * 5**2 * (
    1 / (model_truth_in_model[0, :] ** 2).mean()
    + 1 / (model_truth_in_model[1, :] ** 2).mean()
)

tip_tre_factor = 1 / (12 * num_repetitions) + 1 / 2 * 3**2 * (
    1 / (tip_truth_in_tip[0, :] ** 2).mean() + 1 / (tip_truth_in_tip[1, :] ** 2).mean()
)

tre["aurora_2_model"] = fle["aurora_2_model"] * model_tre_factor
tre["aurora_2_tip"] = fle["aurora_2_tip"] * tip_tre_factor
tre["tip_2_model"] = np.sqrt(tre["aurora_2_tip"] ** 2 + tre["aurora_2_model"] ** 2)

print(model_tre_factor, tip_tre_factor, tre)


if output:
<<<<<<< HEAD
    np.savetxt(Path("../tools/T_aurora_2_model"), T_aurora_2_model, delimiter=",")
    np.savetxt(Path("../tools/T_tip_2_coil"), T_tip_2_coil, delimiter=",")
=======
    np.savetxt(continuum_name.joinpath("tools", "T_aurora_2_model"), T_aurora_2_model, delimiter=",")
    np.savetxt(continuum_name.joinpath("tools", "T_tip_2_coil"), T_tip_2_coil, delimiter=",")
>>>>>>> 76a52845f12c8942810a157345cea541238dbac0
